internal/docstore/firedocstore/native_codec_test.go

```diff
@@ -0,0 +1,153 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firedocstore
+
+import (
+	"context"
+	"io"
+	"net"
+	"testing"
+
+	"cloud.google.com/go/firestore"
+	ts "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// A nativeCodec encodes and decodes structs using the cloud.google.com/go/firestore
+// client. Since that package doesn't export its codec, we have to go behind the
+// scenes and intercept traffic at the gRPC level. We use interceptors to do that. (A
+// mock server would have worked too.)
+
+type nativeCodec struct {
+	client *firestore.Client
+	doc    *pb.Document
+}
+
+func newNativeCodec() (*nativeCodec, error) {
+	// Establish a gRPC server, just so we have a connection to hang the interceptors on.
+	srv := grpc.NewServer()
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		if err := srv.Serve(l); err != nil {
+			panic(err) // we should never get an error because we just connect and stop
+		}
+	}()
+	nc := &nativeCodec{}
+
+	conn, err := grpc.Dial(l.Addr().String(),
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithUnaryInterceptor(nc.interceptUnary),
+		grpc.WithStreamInterceptor(nc.interceptStream))
+	if err != nil {
+		return nil, err
+	}
+	conn.Close()
+	srv.Stop()
+	nc.client, err = firestore.NewClient(context.Background(), "P", option.WithGRPCConn(conn))
+	if err != nil {
+		return nil, err
+	}
+	return nc, nil
+}
+
+// Intercept all unary (non-streaming) RPCs. The only one we should ever get is a Commit, for
+// the Create call in Encode.
+// If this completes successfully, the encoded *pb.Document will be in c.doc.
+func (c *nativeCodec) interceptUnary(_ context.Context, method string, req, res interface{}, _ *grpc.ClientConn, _ grpc.UnaryInvoker, _ ...grpc.CallOption) error {
+	c.doc = req.(*pb.CommitRequest).Writes[0].GetUpdate()
+	cres := res.(*pb.CommitResponse)
+	cres.WriteResults = []*pb.WriteResult{{}}
+	return nil
+}
+
+// Intercept all streaming RPCs. The only one we should ever get is a BatchGet, for the Get
+// call in Decode.
+// Before this is called, c.doc must be set to the *pb.Document to be returned from the call.
+func (c *nativeCodec) interceptStream(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return &clientStream{ctx: ctx, doc: c.doc}, nil
+}
+
+// clientStream is a fake client stream. It returns a single document, then terminates.
+type clientStream struct {
+	ctx context.Context
+	doc *pb.Document
+}
+
+func (cs *clientStream) RecvMsg(m interface{}) error {
+	if cs.doc != nil {
+		cs.doc.CreateTime = &ts.Timestamp{}
```

Comment:
Is it ok to leave all other fields as is? I see there are a few `XXX_` named fields, will those affect the cmp result?

Reply:
1. It's OK to leave the other fields unset; this is the minimum required for the firestore client to complete without an error.

2. I call `cmp.Equal` on a struct I define in the test, not protos, so it should be OK.